### PR TITLE
Plot directive preserve

### PR DESCRIPTION
**Pull request by r-barnes**
## PR Summary

The Sphinx plot directive can be used to automagically generate figures for documentation like so:

    .. plot::

       import matplotlib.pyplot as plt
       import matplotlib.image as mpimg
       import numpy as np
       img = mpimg.imread('_static/stinkbug.png')
       imgplot = plt.imshow(img)

But, if you reorder the figures in the documentation then all the figures may need to be rebuilt. This takes time. The names given to the figures are also fairly meaningless, making them more difficult to index by search engines or to find on a filesystem.

Alternatively, if you are compiling on a limited-resource service like ReadTheDocs, you may wish to build imagery locally to avoid hitting resource limits on the server. Using the new changes allows extensive dynamically generated imagery to be used on services like ReadTheDocs.

### The `:outname:` property

These problems are addressed through two new features in the plot directive. The first is the introduction of the `:outname:` property. It is used like so:

    .. plot::
       :outname: stinkbug_plot

       import matplotlib.pyplot as plt
       import matplotlib.image as mpimg
       import numpy as np
       img = mpimg.imread('_static/stinkbug.png')
       imgplot = plt.imshow(img)

Without `:outname:`, the figure generated above would normally be called, e.g. `docfile3-4-01.png` or something equally mysterious. With `:outname:` the figure generated will instead be named `stinkbug_plot-01.png` or even `stinkbug_plot.png`. This makes it easy to understand which output image is which and, more importantly, uniquely keys output images to code snippets.

### The `plot_preserve_dir` configuration value

Setting the `plot_preserve_dir` configuration value to the name of a directory will cause all images with `:outname:` set to be copied to this directory upon generation.

If an image is already in `plot_preserve_dir` when documentation is being generated, this image is copied to the build directory thereby pre-empting generation and reducing computation time in low-resource environments.

Related issues were raised on ReadTheDocs ([link](https://github.com/rtfd/readthedocs.org/issues/3468)) and on StackOverflow ([link](https://stackoverflow.com/q/48066337/752843)).

## PR Checklist

- [ ] Has Pytest style unit tests
 - It would, but I'm not sure how to go about writing them. Sorry :-/
- [X] Code is PEP 8 compliant
 - Leastwise, nothing new is added to the numerous pre-existing violations. PEP8 adherence makes code difficult to read :-/
- [X] New features are documented, with examples if plot related
- [X] Documentation is sphinx and numpydoc compliant
- [X] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [X] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
:+1: in principle, but the docs need to build cleanly!

File: lib/matplotlib/sphinxext/plot_directive.py [hunk 1]
```diff
@@ -656,6 +674,9 @@ def render_figures(code, code_path, output_dir, output_base, context,
             for format, dpi in formats:
                 try:
                     figman.canvas.figure.savefig(img.filename(format), dpi=dpi)
+                    if config.plot_preserve_dir and outname:
+                      print("Preserving '{0}' into '{1}'".format(img.filename(format), config.plot_preserve_dir))
```
Comment (jklymak) [inline lib/matplotlib/sphinxext/plot_directive.py]: Not reviewing, but I don't think you want `print` in here do you?  I'd do `_log.info` or `_log.debug`?

Comment (r-barnes) [reply lib/matplotlib/sphinxext/plot_directive.py]: Yes, that's probably what I want, but I didn't know about them. I'll update.

Comment (r-barnes) [reply lib/matplotlib/sphinxext/plot_directive.py]: Actually, I don't see `_log` as a variable in the `plot_directive.py` namespace, so I'm not sure how to make that adjustment.

Comment (jklymak) [reply lib/matplotlib/sphinxext/plot_directive.py]: If its not already in this module, then up at the top: 
```python
# ... other imports
import logging

# ... other imports

_log = logging.getLogger(__name__)

```

Comment (r-barnes) [reply lib/matplotlib/sphinxext/plot_directive.py]: Thanks: got that up.

File: lib/matplotlib/sphinxext/plot_directive.py [hunk 2]
```diff
@@ -262,7 +282,8 @@ def setup(app):
                'format': _option_format,
                'context': _option_context,
                'nofigs': directives.flag,
-               'encoding': directives.encoding
+               'encoding': directives.encoding,
+               'outname':        str
```
Comment (has2k1) [inline lib/matplotlib/sphinxext/plot_directive.py]: Too many spaces.

**Comment by r-barnes** [comment]
Eliminated some spaces

**Comment by jklymak** [comment]
@r-barnes sorry this didn't get a second review yet.  If you give it a rebase and ping me I'll give it a quick look and merge on the basis of @tacaswell 's approval.  We are about to branch 3.2, so it'd be nice to get this in

**Comment by timhoffm** [comment]
Wouldn't `plot_preserve_dir` better be named `plot_cache_dir`?

File: lib/matplotlib/sphinxext/plot_directive.py [hunk 3]
```diff
@@ -684,6 +709,20 @@ def run(arguments, content, options, state_machine, state, lineno):
     rst_file = document.attributes['source']
     rst_dir = os.path.dirname(rst_file)
 
+    #Get output name of the images, if the option was provided
+    outname = options.get('outname', '')
+
+    #Ensure that the outname is unique, otherwise copied images will
+    #not be what user expects
+    if outname and outname in _outname_list:
+      raise Exception("The outname '{0}' is not unique!".format(outname))
```
Comment (tacaswell) [inline lib/matplotlib/sphinxext/plot_directive.py]: please use 4 space indentation.

File: lib/matplotlib/sphinxext/plot_directive.py [hunk 4]
```diff
@@ -766,6 +810,15 @@ def run(arguments, content, options, state_machine, state, lineno):
         build_dir_link = build_dir
     source_link = dest_dir_link + '/' + output_base + source_ext
 
+    #If we previously preserved copies of the generated figures
+    #this copies them into the build directory so that they will
+    #not be remade
+    if config.plot_preserve_dir and outname:
+      outfiles = glob.glob(os.path.join(config.plot_preserve_dir, outname) + '*')
+      for of in outfiles:
+        _log.info("Copying preserved copy of '{0}' into '{1}'".format(of, build_dir))
```
Comment (tacaswell) [inline lib/matplotlib/sphinxext/plot_directive.py]: we are py36+ so we can use f-strings :smile:

File: lib/matplotlib/sphinxext/plot_directive.py (hunk 3 above)
Comment (tacaswell) [reply lib/matplotlib/sphinxext/plot_directive.py]: also, can use f-strings.

**Comment by tacaswell** [comment]
sorry, I read the year when this was opened wrong and thought it was from this January, not from 2 years ago.

File: lib/matplotlib/sphinxext/plot_directive.py [hunk 5]
```diff
@@ -129,12 +137,25 @@
 
     plot_template
         Provide a customized template for preparing restructured text.
+
+    plot_preserve_dir
+        Files with outnames are copied to this directory and files in this
+        directory are copied back from into the build directory prior to the
```
Comment (ivanov) [inline lib/matplotlib/sphinxext/plot_directive.py]: `from` or `into` ?

Comment (timhoffm) [reply lib/matplotlib/sphinxext/plot_directive.py]: `into`. Fixed.

**Comment by timhoffm** [comment]
I took the liberty of pushing an update to fix the doc issues.

File: lib/matplotlib/sphinxext/plot_directive.py [hunk 6]
```diff
@@ -519,7 +548,7 @@ def get_plot_formats(config):
 
 def render_figures(code, code_path, output_dir, output_base, context,
                    function_name, config, context_reset=False,
-                   close_figs=False):
+                   close_figs=False, outname=''):
```
Comment (timhoffm) [inline lib/matplotlib/sphinxext/plot_directive.py]: Would go with `outname=None`, which is IMHO a more standard way for "not defined" than an empty string.

File: doc/users/next_whats_new/2018_12_03_sphinx_plot_preserve.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
+Plot Directive `outname` and `plot_preserve_dir`
```
Comment (timhoffm) [inline doc/users/next_whats_new/2018_12_03_sphinx_plot_preserve.rst]: I propose to rename `plot_preserve_dir` to `plot_cache_dir`.

**Comment by dstansby** [comment]
See https://github.com/matplotlib/matplotlib/pull/25091#issuecomment-1408292620 for context on closing.

**Pull request closed without merging**